javascript/ql/src/semmle/javascript/frameworks/ClientRequests.qll

```diff
@@ -62,222 +95,465 @@ module ClientRequest {
+        )
+      )
+    }
   }
 
-  override DataFlow::Node getUrl() { result = url }
+  /**
+   * A model of a URL request made using the `XMLHttpRequest` browser class.
+   *
+   * Note: Prefer to use the `ClientRequest` class as it is more general.
+   */
+  class XMLHttpRequest extends ClientRequest::Range {
+    XMLHttpRequest() {
+      this = DataFlow::globalVarRef("XMLHttpRequest").getAnInstantiation()
+      or
+      // closure shim for XMLHttpRequest
+      this = Closure::moduleImport("goog.net.XmlHttp").getAnInvocation()
+    }
 
-  override DataFlow::Node getHost() { none() }
+    override DataFlow::Node getUrl() { result = getAMethodCall("open").getArgument(1) }
 
-  override DataFlow::Node getADataNode() {
-    exists(string name | name = "set" or name = "send" or name = "query" |
-      result = this.getAChainedMethodCall(name).getAnArgument()
-    )
-  }
-}
+    override DataFlow::Node getHost() { none() }
 
-/**
- * A model of a URL request made using the `XMLHttpRequest` browser class.
- */
-private class XMLHttpRequest extends ClientRequest::Range {
-  XMLHttpRequest() {
-    this = DataFlow::globalVarRef("XMLHttpRequest").getAnInstantiation()
-    or
-    // closure shim for XMLHttpRequest
-    this = Closure::moduleImport("goog.net.XmlHttp").getAnInstantiation()
-  }
+    override DataFlow::Node getADataNode() { result = getAMethodCall("send").getArgument(0) }
 
-  override DataFlow::Node getUrl() { result = getAMethodCall("open").getArgument(1) }
+    private string getExplicitResponseType() {
+      getAPropertyWrite("responseType").getRhs().mayHaveStringValue(result)
+    }
 
-  override DataFlow::Node getHost() { none() }
+    /**
+     * Gets the response type corresponding to the `response` property,
+     * but not the explicitly typed properties, `reponseText` and `responseXML`.
+     */
+    string getAssignedResponseType() {
+      result = getExplicitResponseType()
+      or
+      not exists(getExplicitResponseType()) and
+      result = "text"
+    }
 
-  override DataFlow::Node getADataNode() { result = getAMethodCall("send").getArgument(0) }
-}
+    /** Gets an event listener registered on this XHR object. */
+    DataFlow::FunctionNode getAnEventListener() {
+      result = getAPropertyWrite("on" + any(string s)).getRhs().getAFunctionValue()
+      or
+      result = getAMethodCall("addEventListener").getArgument(1).getAFunctionValue()
+    }
 
-/**
- * A model of a URL request made using the `XhrIo` class from the closure library.
- */
-private class ClosureXhrIoRequest extends ClientRequest::Range {
-  ClosureXhrIoRequest() {
-    exists(DataFlow::SourceNode xhrIo | xhrIo = Closure::moduleImport("goog.net.XhrIo") |
-      this = xhrIo.getAMethodCall("send")
+    /**
+     * Gets a node that refers to this XHR object.
+     *
+     * In particular, this can be the receiver of an event handler.
+     */
+    DataFlow::SourceNode getAnAlias() {
```

Comment:
`private`?

Reply:
I'd like to be able to override it if non-local aliasing is needed.